sorrydb/agents/json_agent.py

```diff
@@ -0,0 +1,137 @@
+
+    Args:
+        output_path: Path to output JSON file
+        output: list of dicts with sorries and proofs
+    """
+    try:
+        with open(output_path, "w") as f:
+            json.dump(output, f, indent=4, cls=SorryJSONEncoder)
+    except Exception as e:
+        logger.error(f"Error saving proofs to {output_path}: {e}")
+        raise
+
+
+class SorryStrategy(Protocol):
+    def prove_sorry(self, repo_path: Path, sorry: Sorry) -> str | None:
+        """ To be implemented by the agent
+        Args:
+            repo_path: Path to the repository
+            sorry: sorry to prove
+
+        Returns:
+            Proof string to replace "sorry" or None if no proof was found
+        """
+        pass
+
+class JsonAgent:
+    def __init__(self, strategy: SorryStrategy, lean_data_path: Path | None = None):
+        self.strategy = strategy
+        self.lean_data_path = lean_data_path
+
+    def _process_sorries(self, local_sorries: list[Sorry], lean_data_dir: Path) -> list[dict]:
+        proofs = []
+        for sorry in local_sorries:
+            # Prepare the repository (clone and checkout)
+            checkout_path = prepare_repository(
+                sorry.repo.remote,
+                sorry.repo.branch,
+                sorry.repo.commit,
+                lean_data_dir,
+            )
+            if not checkout_path:
+                logger.error(f"Failed to prepare repository: {sorry.repo.remote}")
+                raise Exception(f"Failed to prepare repository: {sorry.repo.remote}")
+
+            # Build the Lean project
+            build_lean_project(checkout_path)
+
+            # Attempt to prove the sorry
+            proof_string = self.strategy.prove_sorry(checkout_path, sorry)
+
+            # Verify the proof
+            proof_verified = False
+            if proof_string is not None:
+                proof_verified = verify_proof(
+                    checkout_path,
+                    sorry.repo.lean_version,
+                    sorry.location,
+                    proof_string,
+                )
+
+            # Return pair of sorry and proof
+            if proof_verified:
+                proofs.append({"sorry": sorry, "proof": proof_string})
+            else:
+                proofs.append({"sorry": sorry, "proof": None})
+        return proofs
+    
+    def _process_sorries_wrapper(self, sorries: list[Sorry]) -> list[dict]:
+        with(
+            contextlib.nullcontext(self.lean_data_path)
+            if self.lean_data_path
+            else TemporaryDirectory()
+        ) as data_dir:
+            lean_data_path = Path(data_dir)
+            return self._process_sorries(sorries, lean_data_path)
+        
+    def process_sorries(self, sorry_json_path: Path, proofs_json_path: Path):
+        sorry_data = load_sorry_json(sorry_json_path)
+        sorries = sorry_data["sorries"]
+        remote_urls = set(sorry.repo.remote for sorry in sorries)
+        proofs = []
```

Comment:
If `load_sorry_json` return type is updated to `List[Sorry]`:

```suggestion
        sorries = load_sorry_json(sorry_json_path)
        remote_urls = set(sorry.repo.remote for sorry in sorries)
        proofs = []
```

Reply:
Done